src/main/java/net/neoforged/neoforge/attachment/AttachmentSync.java

```diff
@@ -0,0 +1,248 @@
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.neoforged.bus.api.SubscribeEvent;
+import net.neoforged.fml.common.EventBusSubscriber;
+import net.neoforged.neoforge.common.util.FriendlyByteBufUtil;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import net.neoforged.neoforge.event.level.ChunkWatchEvent;
+import net.neoforged.neoforge.internal.versions.neoforge.NeoForgeVersion;
+import net.neoforged.neoforge.network.connection.ConnectionType;
+import net.neoforged.neoforge.network.payload.SyncAttachmentsPayload;
+import net.neoforged.neoforge.registries.NeoForgeRegistries;
+import net.neoforged.neoforge.registries.RegistryBuilder;
+import net.neoforged.neoforge.registries.callback.AddCallback;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+@ApiStatus.Internal
+@EventBusSubscriber(modid = NeoForgeVersion.MOD_ID)
+public final class AttachmentSync {
+    /**
+     * Contains all entries added to {@link NeoForgeRegistries#ATTACHMENT_TYPES} with a sync handler.
+     * Should never be registered against directly.
+     */
+    public static final Registry<AttachmentType<?>> SYNCED_ATTACHMENT_TYPES = new RegistryBuilder<>(
+            ResourceKey.<AttachmentType<?>>createRegistryKey(
+                    ResourceLocation.fromNamespaceAndPath(NeoForgeVersion.MOD_ID, "synced_attachment_types")))
+                            .sync(true)
+                            .create();
+
+    public static final AddCallback<AttachmentType<?>> ATTACHMENT_TYPE_ADD_CALLBACK = (registry, id, key, value) -> {
+        if (value.syncHandler != null) {
+            Registry.register(SYNCED_ATTACHMENT_TYPES, key.location(), value);
+        }
+    };
+
+    static SyncAttachmentsPayload.Target syncTarget(AttachmentHolder holder) {
+        return switch (holder) {
+            case BlockEntity blockEntity -> new SyncAttachmentsPayload.BlockEntityTarget(blockEntity.getBlockPos());
+            case AttachmentHolder.AsField asField when asField.getExposedHolder() instanceof LevelChunk chunk -> new SyncAttachmentsPayload.ChunkTarget(chunk.getPos());
+            case Entity entity -> new SyncAttachmentsPayload.EntityTarget(entity.getId());
+            case Level ignored -> new SyncAttachmentsPayload.LevelTarget();
+            default -> throw new UnsupportedOperationException("Attachment holder class is not supported: " + holder);
+        };
+    }
+
+    private static <T> void syncUpdate(AttachmentHolder holder, AttachmentType<T> type, List<ServerPlayer> players) {
+        RegistryAccess registryAccess = null;
+        for (var player : players) {
+            if (type.syncHandler.sendToPlayer(holder.getExposedHolder(), player)) {
+                registryAccess = player.registryAccess();
+                break;
+            }
+        }
+        // This also serves as a short-circuit if there are no players to sync data to.
```

Comment:
I would still add an isEmpty check at the start of the method for the player list but I don't care that much, this is probably enough.

Reply:
I don't really see the point, it's not like iterating over an empty list will cause a noticeable performance drop.